File: OSGame/src/quiz.js
// Interactive between-wave quiz in a sky "room".
// - Questions are pre-drawn as a UNIQUE set per quiz (no repeats).
// - No arrow: an instruction is shown centre-screen, SPACE begins.
// - Answer pads live in the sky; walk onto one and dwell to select.
// - Feedback (correct/incorrect + reason) waits for SPACE to continue.
// - Perks are chosen at the END: one pick per correct answer.
import {
  SCREEN_W, SCREEN_H, WORLD_W, QUIZ_PAD_Y, QUIZ_ROOM_TOP, QUIZ_ROOM_BOT, COL, CSS,
} from "./config.js";
import { samplePerks } from "./data/perks.js";

const DWELL = 0.7;
const PAD_W = 210, PAD_H = 90;
const ANSWER_SPACING = 340;
const PERK_SPACING = 430;
const FONT = "Consolas, monospace";

export class Quiz {
  constructor(scene, pool, { stage, total, useCode, boss, learn }) {
    this.scene = scene;
    this.pool = pool;
    this.stage = stage;
    this.boss = boss;
    this.useCode = useCode;
    this.learn = !!learn;   // Learn mode: show a lesson + highlight the answer
    this.total = boss ? 1 : total;

    this.qIndex = 0;
    this.correctCount = 0;
    this.answeredCount = 0;   // questions actually answered (for correct/total)
    this.finished = false;
    this.bigBoostEarned = false;

    this.state = "intro";
    this.pads = [];
    this.originX = WORLD_W / 2;
    this.bossPart = 1;
    this.bossCorrect = 0;
    this.current = null;
    this.lastCorrect = false;
    this.perksRemaining = 0;
    this.questions = [];

    this.uiObjs = [];
    this.room = scene.add.graphics().setScrollFactor(0).setDepth(40);
    this.space = scene.input.keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.SPACE);
    this._drawRoomBand();
  }

  begin(playerSprite) {
    const half = ANSWER_SPACING * 2 + PAD_W;
    this.originX = Math.max(half, Math.min(WORLD_W - half, playerSprite.x));
    if (!this.boss) this.questions = this.pool.drawUnique(this.total, this.stage, this.useCode);
    this._showIntro();
  }

  // ---------------------------------------------------------------- helpers
  _text(x, y, str, size, color, opts = {}) {
    const t = this.scene.add.text(x, y, str, {
      fontFamily: FONT, fontSize: `${size}px`, color,
      align: opts.align || "center",
      wordWrap: opts.wrap ? { width: opts.wrap } : undefined,
      fontStyle: opts.bold ? "bold" : "normal",
    }).setOrigin(opts.originX ?? 0.5, opts.originY ?? 0.5).setScrollFactor(0).setDepth(55);
    this.uiObjs.push(t);
    return t;
  }

  _clearUI() { this.uiObjs.forEach((o) => o.destroy()); this.uiObjs = []; }
  _clearPads() {
    this.pads.forEach((p) => {
      p.rect.destroy(); p.inside.destroy(); p.top.destroy(); p.topBg.destroy(); p.bar.destroy();
    });
    this.pads = [];
  }

  _drawRoomBand() {
    const cam = this.scene.cameras.main;
    const top = QUIZ_ROOM_TOP - cam.scrollY;
    const h = QUIZ_ROOM_BOT - QUIZ_ROOM_TOP;
    this.room.clear();
    this.room.fillStyle(0x1e2240, 0.55).fillRect(0, top, SCREEN_W, h);
    this.room.lineStyle(2, COL.primary, 1);
    this.room.lineBetween(0, top, SCREEN_W, top);
    this.room.lineBetween(0, top + h, SCREEN_W, top + h);
  }

  _centerCard(lines) {
    // lines: [{str,size,color,bold}] rendered stacked, centred
    const panel = this.scene.add.rectangle(SCREEN_W / 2, SCREEN_H / 2, SCREEN_W - 120, 260,
      0x10142a, 0.94).setStrokeStyle(2, COL.primary).setScrollFactor(0).setDepth(54);
    this.uiObjs.push(panel);
    let y = SCREEN_H / 2 - (lines.length - 1) * 26;
    for (const ln of lines) {
      this._text(SCREEN_W / 2, y, ln.str, ln.size, ln.color,
        { wrap: SCREEN_W - 180, bold: ln.bold });
      y += ln.gap ?? 52;
    }
  }

  // ---------------------------------------------------------------- states
  _showIntro() {
    this._clearUI();
    const kind = this.boss ? "BOSS QUIZ" : "QUIZ TIME!";
    this._centerCard([
      { str: kind, size: 36, color: CSS.yellow, bold: true, gap: 60 },
      { str: "Walk UP into the sky room and stand on an answer pad to choose your answer.",
        size: 19, color: CSS.white, gap: 70 },
      { str: "Press SPACE to begin", size: 20, color: CSS.green },
    ]);
    this.state = "intro";
  }

  _loadQuestion() {
    this._clearUI(); this._clearPads();
    if (this.boss) { this._boss = this.pool.getBoss(); this.current = this._boss.part1; }
    else { this.current = this.questions[this.qIndex]; }
    this._beginQuestion();
  }

  // In Learn mode, teach the concept first; in Review mode go straight to it.
  _beginQuestion() {
    this._clearUI(); this._clearPads();
    if (this.learn) this._showLesson();
    else this._presentQuestion();
  }

  _presentQuestion() {
    this._clearUI();
    this._showQuestionUI();
    this._makeAnswerPads();
    this.state = "question";
  }

  _showLesson() {
    const text = this.current.learn || this.current.explanation;
    this._centerCard([
      { str: "LESSON", size: 30, color: CSS.yellow, bold: true, gap: 52 },
      { str: text, size: 18, color: CSS.white, gap: 64 },
      { str: "Press SPACE, then choose the answer that fits the lesson", size: 17, color: CSS.green },
    ]);
    this.state = "lesson";
  }

  _showQuestionUI() {
    const q = this.current;
    const qtext = this._text(SCREEN_W / 2, SCREEN_H - 40, q.q, 20, CSS.white,
      { wrap: SCREEN_W - 100, bold: true, originY: 1 });
    let topY = qtext.y - qtext.height;

    if (q.code) {
      const code = this.scene.add.text(SCREEN_W / 2, topY - 10, q.code, {
        fontFamily: FONT, fontSize: "15px", color: CSS.cyan, align: "left",
      }).setOrigin(0.5, 1).setScrollFactor(0).setDepth(55);
      this.uiObjs.push(code);
      const box = this.scene.add.rectangle(SCREEN_W / 2, code.y - code.height / 2,
        code.width + 28, code.height + 16, 0x0c1018)
        .setStrokeStyle(1, COL.cyan).setScrollFactor(0).setDepth(54);
      this.uiObjs.push(box);
      topY = code.y - code.height - 10;
    }

    const panelTop = topY - 12;
    const panel = this.scene.add.rectangle(SCREEN_W / 2, (panelTop + SCREEN_H) / 2,
      SCREEN_W, SCREEN_H - panelTop, 0x10142a, 0.92).setScrollFactor(0).setDepth(52);
    const line = this.scene.add.rectangle(SCREEN_W / 2, panelTop, SCREEN_W, 2, COL.primary)
      .setScrollFactor(0).setDepth(53);
    this.uiObjs.push(panel, line);
  }

  _makeAnswerPads() {
    const opts = this.current.options;
    const startX = this.originX - ANSWER_SPACING * (opts.length - 1) / 2;
    opts.forEach((text, i) => {
      this._addPad(startX + i * ANSWER_SPACING, QUIZ_PAD_Y,
        String.fromCharCode(65 + i), text, i, COL.primary);
    });
  }

  _makePerkPads(ps) {
    this._clearPads();
    const choices = samplePerks(ps, 3);
    const startX = this.originX - PERK_SPACING * (choices.length - 1) / 2;
    choices.forEach((perk, i) => {
      this._addPad(startX + i * PERK_SPACING, QUIZ_PAD_Y,
        String.fromCharCode(65 + i), `${perk.name}\n${perk.desc}`, perk, perk.color);
    });
  }

  // insideLabel: shown big inside the pad; topText: full text shown ON TOP of the pad.
  // highlight: Learn-mode marker for the correct pad.
  _addPad(x, y, insideLabel, topText, payload, color, highlight = false) {
    const rect = this.scene.add.rectangle(x, y, PAD_W, PAD_H, color)
      .setStrokeStyle(highlight ? 4 : 2, highlight ? COL.yellow : COL.white).setDepth(45);
    const inside = this.scene.add.text(x, y, highlight ? `${insideLabel} *` : insideLabel, {
      fontFamily: FONT, fontSize: "18px", color: CSS.white, align: "center",
      fontStyle: "bold", wordWrap: { width: PAD_W - 16 },
    }).setOrigin(0.5).setDepth(46);
    const top = this.scene.add.text(x, y - PAD_H / 2 - 12, topText, {
      fontFamily: FONT, fontSize: "15px", color: CSS.yellow, align: "center",
      wordWrap: { width: PAD_W + 100 },
    }).setOrigin(0.5, 1).setDepth(47);
    const topBg = this.scene.add.rectangle(x, top.y - top.height / 2,
      top.width + 16, top.height + 8, 0x10142a, 0.92).setStrokeStyle(1, color).setDepth(46);
    const bar = this.scene.add.rectangle(x - PAD_W / 2, y + PAD_H / 2 - 6, 0, 7, COL.yellow)
      .setOrigin(0, 0.5).setDepth(46);
    this.pads.push({ x, y, payload, color, progress: 0, rect, inside, top, topBg, bar });
  }

  // ---------------------------------------------------------------- update
  update(dt, playerSprite, ps) {
    this._drawRoomBand();
    const spacePressed = Phaser.Input.Keyboard.JustDown(this.space);

    if (this.state === "intro") {
      if (spacePressed) this._loadQuestion();
      return;
    }

    if (this.state === "lesson") {
      if (spacePressed) this._presentQuestion();
      return;
    }

    if (this.state === "question" || this.state === "perkselect") {
      let selected = null;
      for (const pad of this.pads) {
        const on = Math.abs(playerSprite.x - pad.x) <= PAD_W / 2 &&
                   Math.abs(playerSprite.y - pad.y) <= PAD_H / 2;
        if (on) { pad.progress += dt; if (pad.progress >= DWELL) selected = pad; }
        else pad.progress = Math.max(0, pad.progress - dt * 2);
        pad.bar.width = PAD_W * Math.min(1, pad.progress / DWELL);
      }
      if (selected) {
        if (this.state === "question") this._resolveAnswer(selected, ps);
        else this._resolvePerk(selected, ps);
      }
      return;
    }

    if (this.state === "qfeedback") {
      if (spacePressed) this._afterQuestion(ps);
    } else if (this.state === "perkfeedback") {
      if (spacePressed) this._afterPerk(ps);
    }
  }

  _resolveAnswer(pad, ps) {
    const correct = pad.payload === this.current.answer;
    this.lastCorrect = correct;
    this.answeredCount += 1;
    if (correct) this.correctCount += 1;
    const expl = this.current.explanation;
    let msg;
    if (this.boss) {
      if (correct) this.bossCorrect += 1;
      if (this.bossPart === 1) msg = correct ? "Correct! Now the reason why..." : "Wrong - but face part 2!";
      else {
        const both = this.bossCorrect === 2;
        this.bigBoostEarned = both;
        msg = both ? "BOTH RIGHT! Massive boost incoming!" : "The boss resists... no boost.";
      }
    } else {
      msg = correct ? "Correct!" : "Not quite.";
    }
    this._showFeedback(msg, expl, correct ? CSS.green : CSS.red);
    this.state = "qfeedback";
  }

  _resolvePerk(pad, ps) {
    ps.applyPerk(pad.payload);
    this._showFeedback(`Gained: ${pad.payload.name}!`, pad.payload.desc, CSS.green);
    this.state = "perkfeedback";
  }

  _showFeedback(msg, expl, color) {
    this._clearUI(); this._clearPads();
    this._centerCard([
      { str: msg, size: 30, color, bold: true, gap: 64 },
      { str: expl, size: 18, color: CSS.white, gap: 70 },
      { str: "Press SPACE to continue", size: 18, color: CSS.yellow },
    ]);
  }

  _afterQuestion(ps) {
    if (this.boss) {
      if (this.bossPart === 1) {
        this.bossPart = 2;
        this.current = this._boss.part2;
        this._beginQuestion();
      } else {
        if (this.bigBoostEarned) this._applyBigBoost(ps);
        this._end();
      }
      return;
    }
    this.qIndex += 1;
    if (this.qIndex < this.total) {
      this._loadQuestion();
    } else {
      this.perksRemaining = this.correctCount;
      if (this.perksRemaining > 0) this._openPerkSelect(ps);
      else this._end();
    }
  }

  _afterPerk(ps) {
    this.perksRemaining -= 1;
    if (this.perksRemaining > 0) this._openPerkSelect(ps);
    else this._end();
  }

  _openPerkSelect(ps) {
    this._clearUI(); this._clearPads();
    this._text(SCREEN_W / 2, SCREEN_H - 44,
      `Choose a perk!  (${this.perksRemaining} left) - step onto a pad`, 20, CSS.green,
      { bold: true, originY: 1 });
    const panel = this.scene.add.rectangle(SCREEN_W / 2, SCREEN_H - 38, SCREEN_W, 48, 0x10142a, 0.92)
      .setScrollFactor(0).setDepth(52);
    this.uiObjs.push(panel);
    this._makePerkPads(ps);
    this.state = "perkselect";
  }

  _applyBigBoost(ps) {
    ps.maxHp += 60; ps.hp = ps.maxHp;
    ps.damage *= 1.5; ps.fireRate *= 0.7; ps.speed *= 1.1;
    if (!ps.abilities.shield) ps.unlockAbility("shield");
  }

  _end() {
    this._clearUI(); this._clearPads();
    this.room.destroy();
    this.finished = true;
  }
}

File: quiz.py
"""Interactive between-wave quiz.

The quiz happens in a 'room' up in the sky. When it starts, the player is guided
UP out of the floor by an on-screen arrow. Answer pads are spread far apart so you
can only stand on one at a time. Walk onto a pad and stand on it (dwell) to select.

Perks are chosen at the END of the quiz: you get one perk pick per question you
answered correctly.
"""
import math
import random
import pygame
from pygame.math import Vector2
from settings import *
from perks import eligible_perks

DWELL_TIME = 0.7             # seconds to stand on a pad to select it
PAD_W, PAD_H = 200, 74
ANSWER_SPACING = 340        # far apart so you can't touch two pads at once
PERK_SPACING = 420


class Pad:
    def __init__(self, x, y, label, payload, color=PRIMARY):
        self.rect = pygame.Rect(0, 0, PAD_W, PAD_H)
        self.rect.center = (x, y)
        self.label = label
        self.payload = payload
        self.color = color
        self.progress = 0.0
        self.detail = ""

    def on(self, player_rect):
        return self.rect.colliderect(player_rect)


class QuizSession:
    def __init__(self, pool, stage, num_questions, use_code=False, boss=False):
        self.pool = pool
        self.stage = stage
        self.boss = boss
        self.use_code = use_code
        self.total = 1 if boss else num_questions
        self.q_index = 0
        self.correct_count = 0
        self.finished = False
        self.big_boost_earned = False

        self.state = "intro"        # intro,question,qfeedback,perkselect,perkfeedback,done
        self.timer = 1.6
        self.pads = []
        self.origin = Vector2(WORLD_W / 2, QUIZ_PAD_Y)
        self.boss_part = 1
        self.boss_correct = 0
        self.current = None
        self.last_correct = False
        self.message = ""
        self.explanation = ""
        self.perks_remaining = 0

    # ------------------------------------------------------------------ setup
    def begin(self, player):
        half = ANSWER_SPACING * 2 + PAD_W  # keep the widest cluster on-map
        ox = max(half, min(WORLD_W - half, player.pos.x))
        self.origin = Vector2(ox, QUIZ_PAD_Y)
        self._load_question()

    def _load_question(self):
        if self.boss:
            self._boss = self.pool.get_boss()
            self.current = self._boss["part1"]
        elif self.use_code and random.random() < 0.6:
            self.current = self.pool.get_code()
        else:
            self.current = self.pool.get_concept(self.stage)
        self._make_answer_pads()
        self.state = "question"

    def _make_answer_pads(self):
        self.pads = []
        opts = self.current["options"]
        n = len(opts)
        start_x = self.origin.x - ANSWER_SPACING * (n - 1) / 2
        for i, text in enumerate(opts):
            pad = Pad(start_x + i * ANSWER_SPACING, QUIZ_PAD_Y, chr(65 + i), i)
            pad.detail = text
            self.pads.append(pad)

    def _make_perk_pads(self, player):
        self.pads = []
        pool = eligible_perks(player)
        choices = random.sample(pool, k=min(3, len(pool)))
        start_x = self.origin.x - PERK_SPACING * (len(choices) - 1) / 2
        for i, perk in enumerate(choices):
            pad = Pad(start_x + i * PERK_SPACING, QUIZ_PAD_Y, perk["name"],
                      perk, color=perk["color"])
            pad.detail = perk["desc"]
            self.pads.append(pad)

    # ----------------------------------------------------------------- update
    def update(self, dt, player):
        player_rect = pygame.Rect(0, 0, 40, 40)
        player_rect.center = player.pos

        if self.state == "intro":
            self.timer -= dt
            if self.timer <= 0:
                self.state = "question"
            return

        if self.state in ("question", "perkselect"):
            selected = None
            for pad in self.pads:
                if pad.on(player_rect):
                    pad.progress += dt
                    if pad.progress >= DWELL_TIME:
                        selected = pad
                else:
                    pad.progress = max(0, pad.progress - dt * 2)
            if selected:
                if self.state == "question":
                    self._resolve_answer(selected, player)
                else:
                    self._resolve_perk(selected, player)

        elif self.state == "qfeedback":
            self.timer -= dt
            if self.timer <= 0:
                self._after_question(player)

        elif self.state == "perkfeedback":
            self.timer -= dt
            if self.timer <= 0:
                self._after_perk(player)

    def _resolve_answer(self, pad, player):
        correct = (pad.payload == self.current["answer"])
        self.last_correct = correct
        self.explanation = self.current["explanation"]
        if self.boss:
            if correct:
                self.boss_correct += 1
            if self.boss_part == 1:
                self.message = ("Correct! Now the reason why..." if correct
                                else "Wrong — but face part 2!")
            else:
                both = self.boss_correct == 2
                self.big_boost_earned = both
                self.message = ("BOTH RIGHT! Massive boost incoming!" if both
                                else "The boss resists... no boost.")
        else:
            if correct:
                self.correct_count += 1
                self.message = "Correct!"
            else:
                self.message = "Not quite."
        self.state = "qfeedback"
        self.timer = 3.4

    def _after_question(self, player):
        if self.boss:
            if self.boss_part == 1:
                self.boss_part = 2
                self.current = self._boss["part2"]
                self._make_answer_pads()
                self.state = "question"
            else:
                if self.big_boost_earned:
                    self._apply_big_boost(player)
                self.finished = True
            return
        # normal quiz: advance through all questions first
        self.q_index += 1
        if self.q_index < self.total:
            self._load_question()
        else:
            # perk phase: one pick per correct answer
            self.perks_remaining = self.correct_count
            if self.perks_remaining > 0:
                self._make_perk_pads(player)
                self.state = "perkselect"
            else:
                self.finished = True

    def _resolve_perk(self, pad, player):
        player.apply_perk(pad.payload)
        self.message = f"Gained: {pad.payload['name']}!"
        self.explanation = pad.payload["desc"]
        self.state = "perkfeedback"
        self.timer = 2.0

    def _after_perk(self, player):
        self.perks_remaining -= 1
        if self.perks_remaining > 0:
            self._make_perk_pads(player)
            self.state = "perkselect"
        else:
            self.finished = True

    def _apply_big_boost(self, player):
        player.max_hp += 60
        player.hp = player.max_hp
        player.damage *= 1.5
        player.fire_rate *= 0.7
        player.speed *= 1.1
        if "shield" not in player.abilities:
            player.unlock_ability("shield")

    # ------------------------------------------------------------------- draw
    def draw(self, surf, cam, font, bigfont, smallfont, player=None):
        overlay = pygame.Surface((SCREEN_W, SCREEN_H), pygame.SRCALPHA)
        overlay.fill((10, 12, 25, 110))
        surf.blit(overlay, (0, 0))

        # draw the sky "room" band behind the pads
        room_top = QUIZ_ROOM_TOP - cam.y
        room_h = (QUIZ_ROOM_BOT - QUIZ_ROOM_TOP)
        room = pygame.Rect(0, room_top, SCREEN_W, room_h)
        room_surf = pygame.Surface((SCREEN_W, room_h), pygame.SRCALPHA)
        room_surf.fill((30, 34, 60, 150))
        surf.blit(room_surf, (0, room_top))
        pygame.draw.line(surf, PRIMARY, (0, room_top), (SCREEN_W, room_top), 2)
        pygame.draw.line(surf, PRIMARY, (0, room_top + room_h),
                         (SCREEN_W, room_top + room_h), 2)

        if self.state == "intro":
            self._panel(surf, bigfont, "QUIZ TIME!", 60, YELLOW)
            self._panel(surf, font,
                        "Follow the arrow UP into the room. Stand on an answer pad to choose.",
                        130, WHITE)
            return

        # question / instruction panel at top (never overlaps HUD; HUD hidden in quiz)
        if self.state in ("question", "qfeedback") and self.current:
            code = self.current.get("code")
            self._question_panel(surf, font, smallfont, self.current["q"], code)
        elif self.state in ("perkselect", "perkfeedback"):
            remaining = self.perks_remaining
            self._panel(surf, font,
                        f"Choose a perk!  ({remaining} pick{'s' if remaining != 1 else ''} left)",
                        40, GREEN)

        # pads
        if self.state in ("question", "perkselect"):
            for pad in self.pads:
                self._draw_pad(surf, cam, pad, font, smallfont)
            if player is not None:
                self._draw_arrow(surf, cam, player)

        # feedback
        if self.state in ("qfeedback", "perkfeedback"):
            col = GREEN if self.last_correct or self.state == "perkfeedback" else RED
            self._panel(surf, bigfont, self.message, 220, col)
            self._wrapped(surf, font, self.explanation, 280, WHITE, SCREEN_W - 160)

        # progress
        if self.boss:
            prog = f"BOSS  Part {self.boss_part}/2"
        elif self.state in ("perkselect", "perkfeedback"):
            prog = "Perk selection"
        else:
            prog = f"Question {min(self.q_index + 1, self.total)}/{self.total}"
        surf.blit(smallfont.render(prog, True, DIM), (20, SCREEN_H - 28))

    # --- arrow guiding the player up to the pads ---
    def _draw_arrow(self, surf, cam, player):
        # target: nearest pad center
        target = min(self.pads, key=lambda p: Vector2(p.rect.center).distance_to(player.pos))
        tc = Vector2(target.rect.center)
        if tc.distance_to(player.pos) < 120:
            return
        px = player.pos.x - cam.x
        py = player.pos.y - cam.y - 40
        d = (tc - player.pos)
        if d.length() == 0:
            return
        d = d.normalize()
        tip = Vector2(px, py) + d * 46
        left = Vector2(px, py) + d.rotate(150) * 22
        right = Vector2(px, py) + d.rotate(-150) * 22
        pulse = 180 + int(70 * math.sin(pygame.time.get_ticks() / 150))
        pygame.draw.polygon(surf, (255, 220, 60), [tip, left, right])
        pygame.draw.polygon(surf, (pulse, pulse, 60), [tip, left, right], 3)

    def _draw_pad(self, surf, cam, pad, font, smallfont):
        r = pad.rect.move(-cam.x, -cam.y)
        pygame.draw.rect(surf, pad.color, r, border_radius=12)
        pygame.draw.rect(surf, WHITE, r, 2, border_radius=12)
        if pad.progress > 0:
            w = int(r.width * min(1, pad.progress / DWELL_TIME))
            pygame.draw.rect(surf, YELLOW, (r.left, r.bottom - 7, w, 7), border_radius=3)
        # label above pad
        lab = font.render(pad.label, True, YELLOW)
        surf.blit(lab, lab.get_rect(center=(r.centerx, r.top - 18)))
        # option/perk text wrapped inside pad
        lines = self._fit(pad.detail, smallfont, r.width - 16)
        total_h = len(lines) * (smallfont.get_height() + 1)
        y0 = r.centery - total_h // 2
        for j, line in enumerate(lines):
            t = smallfont.render(line, True, WHITE)
            surf.blit(t, t.get_rect(center=(r.centerx, y0 + j * (smallfont.get_height() + 1) + 8)))

    # -------------------------------------------------------------- text utils
    def _question_panel(self, surf, font, smallfont, question, code):
        max_w = SCREEN_W - 120
        lines = self._fit(question, font, max_w)
        line_h = font.get_height() + 4
        code_lines = code.split("\n") if code else []
        code_h = (len(code_lines) * (smallfont.get_height() + 2) + 20) if code else 0
        pad_y = 20
        box_h = 24 + len(lines) * line_h + (code_h + 12 if code else 0)
        box = pygame.Rect(60, pad_y, max_w, box_h)
        panel = pygame.Surface(box.size, pygame.SRCALPHA)
        panel.fill((18, 22, 40, 235))
        surf.blit(panel, box.topleft)
        pygame.draw.rect(surf, PRIMARY, box, 2, border_radius=10)
        y = box.top + 12
        for line in lines:
            t = font.render(line, True, WHITE)
            surf.blit(t, t.get_rect(midtop=(box.centerx, y)))
            y += line_h
        if code:
            self._draw_code(surf, smallfont, code_lines, y + 6, box)

    def _draw_code(self, surf, font, code_lines, y, box):
        cw = max(font.size(l)[0] for l in code_lines) + 24
        ch = len(code_lines) * (font.get_height() + 2) + 12
        cx = box.centerx - cw // 2
        pygame.draw.rect(surf, (12, 16, 26), (cx, y, cw, ch), border_radius=8)
        pygame.draw.rect(surf, CYAN, (cx, y, cw, ch), 1, border_radius=8)
        for i, line in enumerate(code_lines):
            t = font.render(line, True, CYAN)
            surf.blit(t, (cx + 12, y + 6 + i * (font.get_height() + 2)))

    def _panel(self, surf, font, text, y, color):
        t = font.render(text, True, color)
        surf.blit(t, t.get_rect(center=(SCREEN_W // 2, y)))

    def _wrapped(self, surf, font, text, y, color, max_w):
        for i, line in enumerate(self._fit(text, font, max_w)):
            t = font.render(line, True, color)
            surf.blit(t, t.get_rect(center=(SCREEN_W // 2, y + i * (font.get_height() + 2))))

    def _fit(self, text, font, max_w):
        words = text.split()
        lines, cur = [], ""
        for w in words:
            test = (cur + " " + w).strip()
            if font.size(test)[0] <= max_w:
                cur = test
            else:
                if cur:
                    lines.append(cur)
                cur = w
        if cur:
            lines.append(cur)
        return lines
